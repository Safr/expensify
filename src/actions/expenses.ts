import { Dispatch, ActionCreator } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { toastr } from 'react-redux-toastr';
import database from '../firebase/firebase';
import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { asyncActionStart, asyncActionFinish, asyncActionError } from './async';
import StoreState, { ExpensesI } from '../types';

export interface InitExpense {
  type: '@@INIT';
}

export interface AddExpense {
  type: ADD_EXPENSE;
  expense: ExpensesI;
}

export interface EditExpense {
  type: EDIT_EXPENSE;
  id: string;
  updates: any;
}

export interface RemoveExpense {
  type: REMOVE_EXPENSE;
  id: string;
}

export interface SetExpenses {
  type: SET_EXPENSES;
  expenses: ExpensesI[];
}

export type ExpenseAction = InitExpense | AddExpense | EditExpense | RemoveExpense | SetExpenses;

export const addExpense: ActionCreator<AddExpense> = expense => ({
  type: ADD_EXPENSE,
  expense,
});

export const startAddExpense = (expenseData = {}): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
  }: any = expenseData;
    const expense = {
      description, note, amount, createdAt,
    };
    return database.ref(`users/${uid}/expenses`).push(expense).then(ref => {
      toastr.success('Success', 'Expense has been added');
      dispatch(addExpense({
        id: ref.key,
        ...expense,
      }));
    });
  };

export const removeExpense: ActionCreator<RemoveExpense> = (id) => ({
  type: REMOVE_EXPENSE,
  id,
});

export const startRemoveExpense = (id: string): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).remove()
    .then(() => {
      toastr.success('Success', 'Expense has been deleted');
      dispatch(removeExpense(id));
    });
  };

export const editExpense: ActionCreator<EditExpense> = (id, updates) => ({
  type: EDIT_EXPENSE,
  id,
  updates,
});

export const startEditExpense = (id: string, updates: ExpensesI): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates)
    .then(() => {
      toastr.success('Success', 'Expense has been edited');
      dispatch(editExpense(id, updates));
    });
  };

export const setExpenses: ActionCreator<SetExpenses> = expenses => ({
  type: SET_EXPENSES,
  expenses,
});

export const startSetExpenses = (): ThunkAction<any, StoreState, null> =>
  (dispatch: Dispatch<any>, getState: () => StoreState) => {
    const { uid } = getState().auth;
    dispatch(asyncActionStart());
    return database.ref(`users/${uid}/expenses`)
    .once('value')
    .then(snapshot => {
      const expenses: ExpensesI[] = [];
      snapshot.forEach((childSnapshot: any) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val(),
        });
      });
      dispatch(setExpenses(expenses));
      toastr.success('Success', 'Expenses have been loaded');
      dispatch(asyncActionFinish());
    })
    .catch(e => dispatch(asyncActionError()));
  };
