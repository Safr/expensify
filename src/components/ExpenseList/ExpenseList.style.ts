import styled from 'styled-components';
import { borderColor, darkGrey, mSize, lSize, offWhite, sSize } from '../../theme/variables';
import media from '../../theme/media';

export const UlList = styled.ul`
  margin: 0;
  padding: 0;
  text-align: left;
`;

export const ListHeader = styled.header`
  display: flex;
  justify-content: space-between;
  padding: ${sSize} ${mSize};
  background: ${offWhite};
  border: 1px solid ${borderColor};
  color: ${darkGrey};
`;

export const P = styled.p`
  margin-top: ${mSize};
  font-style: italic;
  ${media.tablet`
    margin-top: ${lSize};
  `}
`;
