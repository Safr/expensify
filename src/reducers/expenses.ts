import { SET_EXPENSES, ADD_EXPENSE, EDIT_EXPENSE, REMOVE_EXPENSE } from '../actionTypes';
import { ExpenseAction } from '../actions';
import { ExpensesI } from '../types';

const expensesReducer = (state = [], action: ExpenseAction): ExpensesI | ExpensesI[] => {
  switch (action.type) {
    case SET_EXPENSES:
      return [...action.expenses];
    case ADD_EXPENSE:
      return [
        ...state,
        action.expense,
      ];
    case REMOVE_EXPENSE:
      return state.filter(({ id }: {id: string}) => id !== action.id);
    case EDIT_EXPENSE:
      return state.map(expense => (
        expense.id === action.id ? { ...expense, ...action.updates } : expense));
    default:
      return state;
  }
};

export default expensesReducer;

