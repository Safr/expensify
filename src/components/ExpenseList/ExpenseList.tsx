import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../../selectors/expenses';
import ExpenseListItem from '../ExpenseListItem/ExpenseListItem';
import Container from '../../theme/grid';
import StoreState, { ExpensesI } from '../../types';
import { ListHeader, UlList, P } from './ExpenseList.style';

interface Props {
  expenses: ExpensesI[];
}

export const ExpenseList: React.SFC<Props> = ({ expenses }) => (
  <Container>
    <ListHeader>
      <div>Expense</div>
      <div>Amount</div>
    </ListHeader>
    <UlList>
      {
        expenses.length === 0 ?
          <P>No expenses</P>
          :
          expenses.map(expense => <ExpenseListItem key={expense.id} {...expense} />)
      }
    </UlList>
  </Container>
);

const mapStateToProps = (state: StoreState) => ({
  expenses: selectExpenses(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);
