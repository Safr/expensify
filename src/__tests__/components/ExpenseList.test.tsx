import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList/ExpenseList';
import expenses from '../../fixtures/expenses';

describe('ExpenseList Component', () => {
  test('should render ExpenseList correctly with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should render ExpenseList with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
