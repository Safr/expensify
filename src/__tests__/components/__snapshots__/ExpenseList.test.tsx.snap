// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList Component should render ExpenseList correctly with expenses 1`] = `
<styled.div>
  <styled.header>
    <div>
      Expense
    </div>
    <div>
      Amount
    </div>
  </styled.header>
  <styled.ul>
    <ExpenseListItem
      amount={77}
      createdAt={0}
      description="Gum"
      id="1"
      key="1"
      note=""
    />
    <ExpenseListItem
      amount={777}
      createdAt={-345600000}
      description="Rent"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={7777}
      createdAt={345600000}
      description="House"
      id="3"
      key="3"
      note=""
    />
  </styled.ul>
</styled.div>
`;

exports[`ExpenseList Component should render ExpenseList with empty message 1`] = `
<styled.div>
  <styled.header>
    <div>
      Expense
    </div>
    <div>
      Amount
    </div>
  </styled.header>
  <styled.ul>
    <styled.p>
      No expenses
    </styled.p>
  </styled.ul>
</styled.div>
`;
